src/storage/src/hummock/event_handler/hummock_event_handler.rs

```diff
@@ -144,99 +182,116 @@ impl HummockEventHandler {
-            {
-                self.upload_handle_manager
-                    .add_epoch_handle(epoch, once(join_handle));
-            } else {
-                break;
-            }
-        }
+        self.uploader.buffer_tracker()
     }
+}
 
-    fn send_sync_result(&mut self, epoch: HummockEpoch, result: HummockResult<SyncResult>) {
-        if let Some(tx) = self.pending_sync_requests.remove(&epoch) {
-            let _ = tx.send(result).inspect_err(|e| {
-                error!("unable to send sync result. Epoch: {}. Err: {:?}", epoch, e);
-            });
-        } else {
-            panic!("send sync result to non-requested epoch: {}", epoch);
+impl HummockEventHandler {
+    async fn next_event(&mut self) -> Option<Either<UploaderEvent, HummockEvent>> {
+        match select(
+            self.uploader.next_event(),
+            self.hummock_event_rx.recv().boxed(),
+        )
+        .await
+        {
+            Either::Left((event, _)) => Some(Either::Left(event)),
+            Either::Right((event, _)) => event.map(Either::Right),
         }
     }
 }
 
 // Handler for different events
 impl HummockEventHandler {
-    fn handle_epoch_finished(&mut self, epoch: HummockEpoch) {
-        // TODO: in some case we may only need the read guard.
-        let mut local_version_guard = self.local_version_manager.local_version.write();
-        if epoch > local_version_guard.get_max_sync_epoch() {
-            // The finished flush task does not belong to any syncing epoch.
-            return;
-        }
-        let sync_epoch = epoch;
-        let compaction_group_index = local_version_guard
-            .pinned_version()
-            .compaction_group_index();
-        let sync_data = local_version_guard
-            .sync_uncommitted_data
-            .get_mut(&sync_epoch)
-            .expect("should find");
-        match sync_data.stage() {
-            SyncUncommittedDataStage::CheckpointEpochSealed(_) => {
-                let (payload, sync_size) = sync_data.start_syncing();
-                let local_version_manager = self.local_version_manager.clone();
-                let join_handle = tokio::spawn(async move {
-                    let _ = local_version_manager
-                        .run_sync_upload_task(
-                            payload,
-                            compaction_group_index,
-                            sync_size,
-                            sync_epoch,
-                        )
-                        .await
-                        .inspect_err(|e| {
-                            error!("sync upload task failed: {}, err: {:?}", sync_epoch, e);
-                        });
+    fn handle_epoch_synced(
+        &mut self,
+        epoch: HummockEpoch,
+        newly_uploaded_sstables: Vec<StagingSstableInfo>,
+    ) {
+        if !newly_uploaded_sstables.is_empty() {
+            let mut read_version_guard = self.read_version.write();
```

Comment:
Can we prune staging_sst here ? and no need to hold the write lock of read_version and call update with not associated staging_sst  ?  Shall we add some comments and achieve in other PR

Reply:
In the current code, there is only one ReadVersion, so I think there is no need to do the pruning in this PR. Maybe we can do it when we really support sharding.